File: backend_java/src/main/java/br/com/fiap/chronosdtn/model/Bundle.java
package br.com.fiap.chronosdtn.model;

// Por que: Importa as anotações do Jakarta Persistence para persistência ORM de chaves embutidas e CLOBs.
import jakarta.persistence.*;
// Por que: Importa as anotações do Lombok para reduzir código boilerplate de get/set.
import lombok.*;
// Por que: Permite carregar carimbos de data/hora de criação e expiração da fila espacial.
import java.time.LocalDateTime;

// Por que: Declara esta classe como uma entidade JPA gerenciada.
@Entity
// Por que: Associa à tabela de buffers do protocolo DTN (Store-and-Forward).
@Table(name = "TB_CHRONOS_BUNDLE")
// Por que: Lombok - Getters para dados do bundle.
@Getter
// Por que: Lombok - Setters para dados do bundle.
@Setter
// Por que: Lombok - Construtor contendo todos os atributos.
@AllArgsConstructor
// Por que: Lombok - Construtor vazio para o Hibernate.
@NoArgsConstructor
public class Bundle {

    // Por que: Declara o uso de chave composta embutida (@EmbeddedId) para identificação global resiliente.
    @EmbeddedId
    private BundleId id;

    // Por que: Relacionamento de muitos-para-um ligando ao nó remetente físico de origem do pacote.
    @ManyToOne(fetch = FetchType.EAGER)
    // Por que: Mapeia o nó de origem vinculando-se à coluna que já compõe a chave do bundle, exigindo read-only (insert/update false) para evitar conflitos de mapeamento.
    @JoinColumn(name = "ID_NODE_SOURCE", referencedColumnName = "ID_NODE", insertable = false, updatable = false)
    private Node sourceNode;

    // Por que: Relacionamento com o nó receptor final do pacote espacial.
    @ManyToOne(fetch = FetchType.EAGER)
    // Por que: Chave estrangeira que aponta para o ID do nó de destino.
    @JoinColumn(name = "ID_NODE_DEST", nullable = false)
    private Node destNode;

    // Por que: Carga útil serializada contendo os dados da transação bancária em formato JSON.
    @Lob
    @Column(name = "TX_PAYLOAD", nullable = false)
    private String payload;

    // Por que: Hash SHA-256 para checagem rápida de integridade binária do payload.
    @Column(name = "TX_HASH", nullable = false, length = 64)
    private String hash;

    // Por que: Nível de prioridade de entrega espacial (0 = Baixa, 1 = Média, 2 = Alta).
    @Column(name = "NR_PRIORITY", nullable = false)
    private Integer priority;

    // Por que: Estado de trânsito físico do pacote (ex: BUFFERED, IN_TRANSIT, DELIVERED, EXPIRED).
    @Column(name = "ST_TRANSMISSION", nullable = false, length = 20)
    private String transmissionStatus;

    // Por que: Data de empacotamento no gateway de origem.
    @Column(name = "DT_CREATED", nullable = false)
    private LocalDateTime createdTime;

    // Por que: Tempo de vida útil do pacote (TTL) após o qual o bundle expira.
    @Column(name = "DT_EXPIRY", nullable = false)
    private LocalDateTime expiryTime;
}

File: backend_net/ChronosDtnNetApi/Models/Bundle.cs
using System;

namespace ChronosDtnNetApi.Models;

// Por que: Representa o pacote básico (Bundle) do protocolo DTN, contendo dados financeiros serializados e prioridade.
public class Bundle
{
    // Por que: Parte 1 da chave composta. Identificador do nó que gerou o pacote.
    public long SourceNodeId { get; set; }

    // Por que: Instância do nó remetente que realizou o empacotamento original.
    public Node? SourceNode { get; set; }

    // Por que: Parte 2 da chave composta. Sequencial numérico incremental local daquele nó específico.
    public long LocalSequenceId { get; set; }

    // Por que: Identificador estrangeiro do nó destinatário final do pacote espacial.
    public long DestNodeId { get; set; }

    // Por que: Instância do nó destinatário de destino final.
    public Node? DestNode { get; set; }

    // Por que: Dados serializados da transação financeira em formato JSON.
    public string Payload { get; set; } = string.Empty;

    // Por que: Assinatura criptográfica hash SHA-256 para auditoria de integridade física.
    public string Hash { get; set; } = string.Empty;

    // Por que: Prioridade do pacote (0 = Baixa/Telemetria, 1 = Média/Operacional, 2 = Alta/Financeira).
    public int Priority { get; set; }

    // Por que: Estado de trânsito físico do pacote (BUFFERED, IN_TRANSIT, DELIVERED, EXPIRED).
    public string TransmissionStatus { get; set; } = string.Empty;

    // Por que: Data/Hora de criação do pacote no nó de origem.
    public DateTime CreatedTime { get; set; }

    // Por que: Data/Hora de expiração final (TTL) do pacote orbital.
    public DateTime ExpiryTime { get; set; }
}
